### ISSUE TITLE
Should we migrate ValueNode to records in C#

### ISSUE THREAD

**Issue by vitek-karas** [mono/linker#1924]
It seems that we try to keep the ValueNode and derived classes immutable, so maybe we should use the records feature to simplify the code.

**Comment by marek-safar** [comment]
I think it's worth the shot to see how the code will look like

**Comment by agocke** [comment]
This was one of my preferred use cases for records, so I'd be interested in how it goes 😄

**Comment by am11** [comment]
Records are useful when we need to make use of the boilerplate code which roslyn generates (overloaded operators, ToString/GetHashCode/Equals etc.).
If codegen is a concern here, then records are <kbd>*highly*</kbd> inefficient compared to classes or struct. e.g. given:

```diff
C c = new();
c.X = 1;
c.Y = 2;

- public record C
+ pubic class C
{
    public int X;
    public int Y;
}
```

here is an IL diff https://www.diffchecker.com/pGvzZ1q0 (443 lines of IL with record, vs. 27 lines with class), the native codegen also suffers accordingly: https://www.diffchecker.com/zmKhxny2 (286 lines of native asm with record vs. 14 lines with class).

This is one reason why `record`s haven't gained popularity in the runtime repo. :)

**Comment by MichalStrehovsky** [comment]
> If codegen is a concern here, then records are <kbd>*highly*</kbd> inefficient compared to classes or struct. e.g. given

Oh wow. Now I wonder if we will need an IL Linker feature to do something with all that cruft once this takes off. This is a lot more overhead than I expected.

**Comment by agocke** [comment]
This is just the structural implementation of IEquatable and ToString. To trim it we'll probably need to prove that a given record is never passed to a method that calls `Equals`.

Alternatively, if you have a tree of nodes and you never plan to use structual equatability you could just override the implementation in your base.

**Comment by MichalStrehovsky** [comment]
Yeah, that's pretty difficult to prove statically that ToString/Equals isn't used. We sometimes know what is the exact type ToString is called on, but as soon as there's one ToString call where we aren't sure what the instance is, it could be anything that got allocated. I don't think we could trim these even for a hello world; this effectively roots all the fields and their types. Probably along with their interfaces because the field types are used as generic arguments. We should definitely avoid this in our libraries.

For IL Linker, we should mostly measure perf - we don't care about the trimmability. The existing structure of ValueNode appears to be optimized for quick comparisons (e.g. I assume it's why `Kind` is there in the first place and is not virtual). Since this was originally written by Scott, I would tend to believe there were good reasons for the structure.
